server/services/store/storetests/workspaces.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package storetests
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mattermost/focalboard/server/model"
+	"github.com/mattermost/focalboard/server/utils"
+
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/mattermost/focalboard/server/services/store"
+)
+
+func StoreTestWorkspaceStore(t *testing.T, setup func(t *testing.T) (store.Store, func())) {
+	t.Run("UpsertWorkspaceSignupToken", func(t *testing.T) {
+		store, tearDown := setup(t)
+		defer tearDown()
+		testUpsertWorkspaceSignupToken(t, store)
+	})
+
+	t.Run("UpsertWorkspaceSettings", func(t *testing.T) {
+		store, tearDown := setup(t)
+		defer tearDown()
+		testUpsertWorkspaceSettings(t, store)
+	})
+
+	t.Run("GetWorkspaceCount", func(t *testing.T) {
+		store, tearDown := setup(t)
+		defer tearDown()
+		testGetWorkspaceCount(t, store)
+	})
+}
+
+func testUpsertWorkspaceSignupToken(t *testing.T, store store.Store) {
+	t.Run("Insert and update workspace with signup token", func(t *testing.T) {
+		workspaceID := "0"
+		workspace := &model.Workspace{
+			ID:          workspaceID,
+			SignupToken: utils.CreateGUID(),
+		}
+
+		// insert
+		err := store.UpsertWorkspaceSignupToken(*workspace)
+		require.NoError(t, err)
+
+		got, err := store.GetWorkspace(workspaceID)
+		require.NoError(t, err)
+		require.Equal(t, workspace.ID, got.ID)
+		require.Equal(t, workspace.SignupToken, got.SignupToken)
+
+		// update signup token
+		workspace.SignupToken = utils.CreateGUID()
+		err = store.UpsertWorkspaceSignupToken(*workspace)
+		require.NoError(t, err)
+
+		got, err = store.GetWorkspace(workspaceID)
+		require.NoError(t, err)
+		require.Equal(t, workspace.ID, got.ID)
+		require.Equal(t, workspace.SignupToken, got.SignupToken)
+	})
+}
+
+func testUpsertWorkspaceSettings(t *testing.T, store store.Store) {
+	t.Run("Insert and update workspace with settings", func(t *testing.T) {
+		workspaceID := "0"
+		workspace := &model.Workspace{
+			ID: workspaceID,
+			Settings: map[string]interface{}{
+				"field1": "A",
+			},
+		}
+
+		// insert
+		err := store.UpsertWorkspaceSettings(*workspace)
+		// FIXME: insert workspace by store.UpsertWorkspaceSettings will failed
```

Comment:
@jespino is this expected?

Reply:
No, it is not expected, you should be able to insert the workspace with the signup_token or make the signup_token not mandatory, clearly this is a bug.

Reply:
I think this should create a random signup_token on create, but keep it untouched on update.